File: 205-isomorphic-strings.cpp
//https://leetcode.com/problems/isomorphic-strings
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if(s.size() != t.size())
            return false;
        if(s == t)
            return true;
        int index = 0;
        unordered_map<char, char> m;
        unordered_map<char, char> m_rev;
        while(index < t.size()){
            if(m.find(s[index]) == m.end() && m_rev.find(t[index]) == m_rev.end()){
                m[s[index]] = t[index];
                m_rev[t[index]] = s[index];
            }else if(m.find(s[index]) == m.end() && m_rev.find(t[index]) != m_rev.end()){
                return false;
            }else if(m.find(s[index]) != m.end() && m_rev.find(t[index]) == m_rev.end()){
                return false;
            }else{
                if(m[s[index]] != t[index] || m_rev[t[index]] != s[index])
                    return false;
            }
            index++;
        }
        return true;
    }
};

// "badc"
// "baba"

File: 205-isomorphic-strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        m_s_t = {}
        m_t_s = {}

        for char1, char2 in zip(s, t):
            if (char1 in m_s_t and m_s_t[char1] != char2) or (char2 in m_t_s and m_t_s[char2] != char1):
                return False
            else:
                m_s_t[char1] = char2
                m_t_s[char2] = char1

        return True

"""
"bbbaaaba"
"aaabbbba"
"badc"
"baba"
"""
